### PR TITLE
Fix selection issue

### LINKED ISSUE #638: Text inserted at incorrect position
## Steps to reproduce
- Type four or five lines of text
- Apply the following formatting
  - Line 1: bold
  - Line 2: list
  - Line 3: bold
  - Line 4: list
  - Line 5: list
- Move the cursor to the middle of line 5
- Type any text
- _Note the text is inserted at the wrong position_

## Details
- Platform: Web (Desktop) - not tested on other platforms
- App version: Element Nightly version: 2023032601

## Screen recording

[Screencast from 29-03-23 09:40:16.webm](https://user-images.githubusercontent.com/4940864/228477525-06bd40b9-8f5e-4647-a915-e9fae2d6d624.webm)

### PR DESCRIPTION
**Pull request by alunturner**
Fixes #638 

This has probably been around a little while, since we introduced paragraphs. It was the particular edge case where you have a formatting node in a paragraph node that's a sibling of a list.

Have made the fix and added a test to cover both list cases.

### CODE REVIEW

File: platforms/web/lib/dom.ts [hunk 1]
```diff
@@ -538,6 +538,13 @@ function isNodeRequiringExtraOffset(node: Node) {
...
     // do not add an extra offset for the ol/ul tag, only for the list items
     // themselves
+
+    // with the paragraph refactor, we could end up in a situation where we have
+    // a p tag containing formatted text next to a list, so if this is the case,
+    // we do want to add an extra offset _if_ the list has any children
+    if (node.nodeName === 'UL' || node.nodeName === 'OL') {
+        return node.hasChildNodes();
+    }
     return EXTRA_OFFSET_NODE_NAMES.includes(node.nodeName || '');
```
Comment (jonnyandrew) [inline platforms/web/lib/dom.ts:548]: Could `EXTRA_OFFSET_NODE_NAMES` be a bit misleading now as it doesn't contain `UL` or `OL`?

**Review by jonnyandrew — APPROVED** [review]
Looks good to me!

I see you've added unit tests for the offset checking function but would it be good to have one for the whole behaviour that is fixed?

Something like
- Given some affected HTML
- When cursor is placed at an affected position and type
- Then the text is inserted at the correct position

File: platforms/web/lib/dom.ts (hunk 1 above)
Comment (alunturner) [reply platforms/web/lib/dom.ts:548]: I think that's a fair point. I'll have a look at the comments and naming in this section and see if they can be improved.

Comment (alunturner) [reply platforms/web/lib/dom.ts:542]: Thanks for the comment, rereading this I've actually realised I could remove this special case and add the list nodes to the `EXTRA_OFFSET_NODE_NAMES` array. Much clearerf.
